compiler-plugin/src/main/kotlin/com/google/devtools/ksp/processing/impl/ResolverImpl.kt

```diff
@@ -453,6 +461,66 @@ class ResolverImpl(
         return KSTypeArgumentLiteImpl.getCached(typeRef, variance)
     }
 
+    override fun asMemberOf(
+        property: KSPropertyDeclaration,
+        containing: KSType
+    ): KSType {
+        val propertyDeclaredIn = property.closestClassDeclaration()
+            ?: throw IllegalArgumentException("Cannot call asMemberOf with a property that is " +
+                "not declared in a class or an interface")
+        val declaration = resolvePropertyDeclaration(property)
+        if (declaration != null && containing is KSTypeImpl && !containing.isError) {
+            if (!containing.kotlinType.isSubtypeOf(propertyDeclaredIn)) {
+                throw IllegalArgumentException(
+                    "$containing is not a sub type of the class/interface that contains `$property` ($propertyDeclaredIn)"
+                )
+            }
+            val typeSubstitutor = containing.kotlinType.createTypeSubstitutor()
+            val substituted = declaration.substitute(typeSubstitutor) as? ValueDescriptor
+            substituted?.let {
+                return KSTypeImpl.getCached(substituted.type)
+            }
+        }
+        // if substitution fails, fallback to the type from the property
+        return KSErrorType
+    }
+
+    override fun asMemberOf(
+        function: KSFunctionDeclaration,
+        containing: KSType
+    ): KSFunction {
+        val functionDeclaredIn = function.closestClassDeclaration()
+            ?: throw IllegalArgumentException("Cannot call asMemberOf with a function that is " +
+                "not declared in a class or an interface")
+        val declaration = resolveFunctionDeclaration(function)
+        if (declaration != null && containing is KSTypeImpl && !containing.isError) {
+            if (!containing.kotlinType.isSubtypeOf(functionDeclaredIn)) {
+                throw IllegalArgumentException(
+                    "$containing is not a sub type of the class/interface that contains " +
+                        "`$function` ($functionDeclaredIn)"
+                )
+            }
+            val typeSubstitutor = containing.kotlinType.createTypeSubstitutor()
+            val substituted = declaration.substitute(typeSubstitutor)
+            return KSFunctionImpl(substituted)
```

Comment:
Can you cache the creation of `KSFunctionImpl` like [this](https://github.com/google/ksp/blob/master/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/binary/KSAnnotationDescriptorImpl.kt#L41)?

Reply:
i'm not sure if this is worth caching, the value is quite specific to an owner + declaration.
But even if I wanted to cache that, substitute returns `CallableDescriptor` which does not seem to enforse implementing proper `equals`. 

I've actually tried to run substitution twice between a `List<String>` type and `List#get` and they both returned different descriptors.

![image](https://user-images.githubusercontent.com/89202/96520425-ea5c3980-1223-11eb-9b35-d2e4fb906a45.png)

I can try to cache in the resolverImpl's `asMemberOf` with a composite key of `KSFunctionDeclaration,KSType` though again, i'm not sure how useful that cache would be.

Reply:
> with a composite key of `KSFunctionDeclaration,KSType`
Sounds good to me. Besides performance*, like what you said, it comes with proper `equals`.

* `asMemberOf` is inherently expensive. I'm a little bit surprised that the substituter returned different instances when called twice, which means that some of the computations are redundant.

Reply:
i implemented caching, though i don't have KSFUnction#equals, I guess probably should implement it as well?
Also, might be better to do it as a followup to get this CL in.